src/ClientIp.php

```diff
@@ -107,33 +107,70 @@ public function process(ServerRequestInterface $request, DelegateInterface $dele
      */
     private function getIp(ServerRequestInterface $request)
     {
-        if ($this->remote) {
-            $ip = file_get_contents('http://ipecho.net/plain');
+        if ($ip = $this->getRemoteIp()) {
+            return $ip;
+        }
 
-            if (self::isValid($ip)) {
+        if ($ip = $this->getProxyIp($request)) {
+            if (!in_array($ip, $this->proxyIps)) {
                 return $ip;
             }
         }
 
-        $ip = null;
-        $server = $request->getServerParams();
-
-        if (!empty($server['REMOTE_ADDR']) && self::isValid($server['REMOTE_ADDR'])) {
-            $ip = $server['REMOTE_ADDR'];
+        if ($ip = $this->getLocalIp($request)) {
+            if (!in_array($ip, $this->proxyIps)) {
+                return $ip;
+            }
```

Comment:
Ah, there's a subtle difference that I missed that the tests are not accounting for:

The local IP **has** to be checked against proxy IPs if the proxy IP list is defined. Otherwise something weird is happening. Got it.